### PR TITLE
Extract SignallingWebServer matchmaker and player classes

### PR DESCRIPTION
**Pull request by gunsha**
Extract SignallingWebServer matchmaker socket into it's own class to help make the code more readable
relates to #10

### CODE REVIEW

**Comment by gunsha** [comment]
@lukehb rebased the repo and signed the commit

**Comment by lukehb** [comment]
@Belchy06 This bodes well for the refactor we intend to do in 5.3 anyway. Can I get you to review and QA against multi-streamer and SFU feature set to ensure no regressions?

**Comment by gunsha** [comment]
@lukehb @Belchy06 do you guys have any automation test set for this or is all done manually?

**Comment by Belchy06** [comment]
Unfortunately it's all still manual QA but automated tests are planned

**Comment by lukehb** [comment]
@MWillWallT If Will is swamped, can you also QA this PR to ensure things are still working, particularly multi-streamer and SFU feature set.

File: SignallingWebServer/cirrus.js [hunk 1]
```diff
@@ -10,8 +10,12 @@ const path = require('path');
 const querystring = require('querystring');
 const bodyParser = require('body-parser');
 const logging = require('./modules/logging.js');
+const WebSocket = require('ws');
```
Comment (Belchy06) [inline SignallingWebServer/cirrus.js:13]: Need to either remove this line or L317 as there's now 2 definitions

File: SignallingWebServer/modules/player.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
+module.exports = class Player {
+	constructor(id, ws, type, browserSendOffer) {
+		this.id = id;
+		this.ws = ws;
+		this.type = type;
+		this.browserSendOffer = browserSendOffer;
+	}
+
+	subscribe(streamerId) {
+		if (!streamers.has(streamerId)) {
```
Comment (Belchy06) [inline SignallingWebServer/modules/player.js:15]: The player class doesn't know about this variable which is global to cirrus.js

File: SignallingWebServer/modules/player.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		this.streamerId = streamerId;
+		const msg = { type: 'playerConnected', playerId: this.id, dataChannel: true, sfu: this.type == PlayerType.SFU, sendOffer: !this.browserSendOffer };
+		logOutgoing(this.streamerId, msg);
+		this.sendFrom(msg);
+	}
+
+	unsubscribe() {
+		if (this.streamerId && streamers.has(this.streamerId)) {
+			const msg = { type: 'playerDisconnected', playerId: this.id };
+			logOutgoing(this.streamerId, msg);
```
Comment (Belchy06) [inline SignallingWebServer/modules/player.js]: logOutgoing will undefined as that method is defined in cirrus.js

File: SignallingWebServer/modules/matchmaker.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+		this.#sendMessage({ type: 'clientConnected' });
+	}
+
+	// The Matchmaker is interested when nobody is connected to a Cirrus server
+	// because then it can re-direct clients to this re-cycled Cirrus server.
+	sendPlayerDisconnected(){
+		this.#sendMessage({ type: 'clientDisconnected' });
+	}
+
+	#sendMessage(message){
```
Comment (Belchy06) [inline SignallingWebServer/modules/matchmaker.js:113]: Probably don't want our methods containing special characters

**Review by Belchy06 — CHANGES_REQUESTED** [review]

File: SignallingWebServer/modules/matchmaker.js (hunk 4 above)
Comment (gunsha) [reply SignallingWebServer/modules/matchmaker.js:113]: any reason for this? that's just the way of declaring private members according to [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields)

**Comment by lukehb** [comment]
@gunsha Due to other changes landing, this PR has conflicts that need to be resolved now. Can you do this?

**Comment by gunsha** [comment]
@lukehb pushed some changes to resolve the conflict

**Comment by Belchy06** [comment]
Still getting conflicts. You'll most likely just need to bring your fork up to date as it's 34 commits behind. Once your fork is synced you can rebase your changes and be good to go from there.

**Comment by gunsha** [comment]
@Belchy06 @lukehb repo is synced now 👌

File: SignallingWebServer/package.json [hunk 5]
```diff
@@ -30,5 +31,8 @@
     "ws": "^7.1.2",
     "y18n": "^5.0.5",
     "yargs": "^15.3.0"
+  },
+  "devDependencies": {
+    "jest": "^29.4.3"
```
Comment (Belchy06) [inline SignallingWebServer/package.json]: Can you just change this to be `"jest": "27.5.1"`. Version 29 requires >= node 16.10 but we explicitly use 16.4.2 in the install scripts.

I've verified testing the suite with v27.5.1 still passes.

**Review by Belchy06 — COMMENTED** [review]
These changes are non-functional and actually crash the signalling server when a player connects.

Do you have the ability to test these changes against a running Pixel Streaming application? If so, I'd implore you to try running the updated signalling server and testing both player and SFU connection.

File: SignallingWebServer/package.json (hunk 5 above)
Comment (gunsha) [reply SignallingWebServer/package.json]: changed the jest version

**Comment by gunsha** [comment]
@Belchy06 tested against SFU and browser, now it's working

**Comment by lukehb** [comment]
@Belchy06 or @MWillWallT to test this PR - particularly with multi-streamer feature and SFU

File: SignallingWebServer/cirrus.js [hunk 6]
```diff
@@ -455,12 +385,22 @@ streamerMessageHandlers.set('offer', forwardStreamerMessageToPlayer);
...
+	matchmaker.connect();
+	matchmaker.registerKeepAlive(matchmakerKeepAliveInterval);
+}
 
 console.logColor(logging.Green, `WebSocket listening for Streamer connections on :${streamerPort}`)
 let streamerServer = new WebSocket.Server({ port: streamerPort, backlog: 1 });
 streamerServer.on('connection', function (ws, req) {
 	console.logColor(logging.Green, `Streamer connected: ${req.connection.remoteAddress}`);
-	sendStreamerConnectedToMatchmaker();
+	matchmaker?.setStreamerReadyState(streamer.readyState);
```
Comment (Belchy06) [inline SignallingWebServer/cirrus.js]: You're attempting to use the streamer here but it isn't declared until line 405. This causes a crash in its current state

**Comment by Belchy06** [comment]
Unfortunately, due to the sections this PR is attempting to modify, it needs a very in depth QA process.

We've recently moved the Unreal Engine docs for Pixel Streaming to this repo. I'd suggest trying to stream with a matchmaker as well, the docs should explain how to set that up. 

Let us know if you need any help.

Cheers

**Comment by lukehb** [comment]
@gunsha In addition to @Belchy06  comment, master has changed again due to fixing some security issues last week. I will need you to resolve conflicts and address @Belchy06  comments before I can merge this one.

As @Belchy06 said, let us know if you need any help from our side. This change is a good one and I would like to get it merged assuming we can ensure everything is still working as before.

**Comment by gunsha** [comment]
Hey @lukehb I'll be checking this tomorrow, I should be able to run the whole thing completely, I'll let you know guys if I need anything from your side 👌

**Comment by gunsha** [comment]
@lukehb tested again with and without matchmaker and seems to be working fine, couldn't try again with SFU, moved to an M2 mac an seems that the mediasoup_prebuilt is not working

**Comment by lukehb** [comment]
@Belchy06 or @MWillWallT Can either of you test this on @gunsha behalf.

@gunsha I see there are once again conflicts that need to be resolved.

**Comment by gunsha** [comment]
@lukehb merged with master

**Comment by Belchy06** [comment]
This one looks good to go @gunsha.

We're currently winding down active development of the master branch in preparation for the 5.2 launch, meaning we're only merging critical bug fixes.

We'll open back up to features once 5.2 has launched.

**Comment by Belchy06** [comment]
Heya @gunsha,

This has been cherry picked as part of a larger commit.

Your changes can be seen here: https://github.com/EpicGames/PixelStreamingInfrastructure/commit/c89193e46d9071de40507f0a522e34b0b6c7f462

Thank you!

**Pull request closed without merging**